test/com/ableton/JenkinsMocks.groovy

```diff
@@ -17,4 +17,74 @@ class JenkinsMocks {
       return System.properties['user.dir']
     }
   }
+
+  /**
+   * Simple container for holding mock script output.
+   * @see JenkinsMocks#addShMock
+   */
+  class MockScriptOutput {
+    String stdout
+    int exitValue
+
+    MockScriptOutput(String stdout, int exitValue) {
+      this.stdout = stdout
+      this.exitValue = exitValue
+    }
+  }
+
+  /** Holds configured mock output values for the `sh` command. */
+  static Map<String, MockScriptOutput> mockScripts = [:]
+
+  /**
+   * Configure mock output for the `sh` command. This function should be called before
+   * attempting to call JenkinsMocks.sh.
+   * @see JenkinsMocks#sh
+   * @param script Script command to mock.
+   * @param stdout Standard output text to return for the given command.
+   * @param exitValue Exit value for the command (default: 0).
+   * @return
+   */
+  static addShMock(String script, String stdout, int exitValue = 0) {
+    mockScripts[script] = new MockScriptOutput(null, stdout, exitValue)
+  }
+
+  static Closure sh = { args ->
+    String script = null
+    def returnStdout = false
+    def returnStatus = false
+
+    // The `sh` function can be called with either a string, or a map of key/value pairs.
+    if (args instanceof String || args instanceof GString) {
+      script = args
+    } else if (args instanceof Map) {
+      script = args['script']
+      returnStatus = args['returnStatus'] ?: false
+      returnStdout = args['returnStdout'] ?: false
+      if (returnStatus && returnStdout) {
+        throw new IllegalArgumentException(
+          'returnStatus and returnStdout are mutually exclusive options')
+      }
+    }
+    assert script
+
+    MockScriptOutput output = mockScripts[script]
+    if (!output) {
+      throw new IllegalArgumentException("No mock output configured for script call " +
+        "'${script}', did you forget to call JenkinsMocks.addShMock()?")
+    }
+    if (!returnStdout) {
+      println output.stdout
+    }
+    if (output.exitValue != 0) {
+      throw new Exception('Mock script failed with error code: ' + output.exitValue)
```

Comment:
If we throw when `output.exitValue != 0`, then we can always return 0 at line 85. So I think we shouldn't throw, right?

Reply:
This is meant to emulate Jenkins' behavior for `sh`. In a Jenkins pipeline, `sh` will throw if the exit code is anything other than 0.

Reply:
Then I don't understand how `returnStatus` works...